kong/plugins/bot-detection/handler.lua

```diff
@@ -0,0 +1,71 @@
+local BasePlugin = require "kong.plugins.base_plugin"
+local reg = require "rex_pcre"
+local responses = require "kong.tools.responses"
+local rules = require "kong.plugins.bot-detection.rules"
+local stringy = require "stringy"
+local bot_cache = require "kong.plugins.bot-detection.cache"
+
+local ipairs = ipairs
+
+local BotDetectionHandler = BasePlugin:extend()
+
+BotDetectionHandler.PRIORITY = 2500
+
+local ALLOW = 1
+local BLOCK = 2
+
+function BotDetectionHandler:new()
+  BotDetectionHandler.super.new(self, "bot-detection")
+end
+
+function BotDetectionHandler:access(conf)
+  BotDetectionHandler.super.access(self)
+
+  local user_agent = ngx.req.get_headers()["user-agent"]
```

Comment:
> Reading/Parsing bodies (the utility function should read the temp files if the bodies are too big)

Yes to most of it, but also regarding this, we should have some sort of API available to plugins, asking Kong core to read the body. Right now, plugins independently read and buffer the body each on their own, instead of asking Kong to do it once for them.